pkg/ingester/metadata_map.go

```diff
@@ -0,0 +1,153 @@
+package ingester
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+)
+
+type metadataMap struct {
+	mtx              sync.Mutex
+	metricToMetadata map[string]*memoryMetadataSet
+}
+
+func newMetadataMap() *metadataMap {
+	return &metadataMap{
+		metricToMetadata: map[string]*memoryMetadataSet{},
+	}
+}
+
+func (mm *metadataMap) purge(max time.Duration) int {
```

Comment:
Functions calling `model.Now()` are hard to test. This specific function can instead be passed "cut-off" time parameter, and entries older than cut-off time are removed. (Same for `set.purge`)

Reply:
I agree that mocking time in Go is a hard thing. However, these are never tested directly, just indirectly through ingesters so that wasn't really a factor. That being said, I think the deadline concept makes it easier to reason about so I have changed to that. Thanks!